File: FruitWholesale.Client/src/app/features/employee-work-log/employee-loan.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { EmployeeLoanHistoryRow, EmployeeLoanRepayment, EmployeeLoanSummaryRow, SaveEmployeeLoanRepayment } from '../../core/models/master-data.model';

@Injectable({ providedIn: 'root' })
export class EmployeeLoanService {
  private readonly baseUrl = `${environment.apiUrl}/employeeloan`;

  constructor(private readonly http: HttpClient) {}

  getSummary(): Observable<EmployeeLoanSummaryRow[]> {
    return this.http.get<EmployeeLoanSummaryRow[]>(`${this.baseUrl}/summary`);
  }

  getHistory(employeeId: number): Observable<EmployeeLoanHistoryRow[]> {
    return this.http.get<EmployeeLoanHistoryRow[]>(`${this.baseUrl}/${employeeId}/history`);
  }

  createRepayment(dto: SaveEmployeeLoanRepayment): Observable<EmployeeLoanRepayment> {
    return this.http.post<EmployeeLoanRepayment>(`${this.baseUrl}/repayments`, dto);
  }

  updateRepayment(id: number, dto: SaveEmployeeLoanRepayment): Observable<EmployeeLoanRepayment> {
    return this.http.put<EmployeeLoanRepayment>(`${this.baseUrl}/repayments/${id}`, dto);
  }

  deleteRepayment(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/repayments/${id}`);
  }
}

File: FruitWholesale.Mobile/lib/features/employee_loan/employee_loan_service.dart
import '../../core/api/api_client.dart';
import 'employee_loan_models.dart';

class EmployeeLoanService {
  final ApiClient _api;

  EmployeeLoanService(this._api);

  Future<List<EmployeeLoanSummaryRow>> getSummary() async {
    final json = await _api.get('/employeeloan/summary');
    return (json as List).map((e) => EmployeeLoanSummaryRow.fromJson(e as Map<String, dynamic>)).toList();
  }

  Future<List<EmployeeLoanHistoryRow>> getHistory(int employeeId) async {
    final json = await _api.get('/employeeloan/$employeeId/history');
    return (json as List).map((e) => EmployeeLoanHistoryRow.fromJson(e as Map<String, dynamic>)).toList();
  }

  Future<void> createRepayment(EmployeeLoanRepayment repayment) async {
    await _api.post('/employeeloan/repayments', body: repayment.toSaveJson());
  }

  Future<void> deleteRepayment(int id) async {
    await _api.delete('/employeeloan/repayments/$id');
  }

  Future<void> applyAdjustment(int employeeId, {required double amount, required bool isIncrease, required String narration}) async {
    await _api.post('/employeeloan/$employeeId/adjustments', body: {
      'amount': amount,
      'isIncrease': isIncrease,
      'narration': narration,
    });
  }

  Future<void> deleteAdjustment(int id) async {
    await _api.delete('/employeeloan/adjustments/$id');
  }
}
